### PR TITLE
sql: Add pg_constraint table to pg_catalog

### PR DESCRIPTION
**Pull request by nvb**
The pg_constraint table provides access to check, primary key,
unique, foreign key, and exclusion constraints on tables.

This is the last `pg_catalog` table needed for GORM support. GORM does rely on the
functionality described in #9990, but it is possible to use the ORM almost completely
without support for OID casts. 

@dt the fk stuff was the trickiest part about this change, primarily because I 
havent dealt much with that code. Do you mind double checking that part specifically?

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9991)

### CODE REVIEW

**Comment by knz** [comment]
:lgtm:

With a grain of salt.

At a high level I cannot see anything wrong with this code.
Looking more closely I as a reviewer hit two issues. First I am lacking detailed references to the pg docs above each table generator, whereby I could check that the code is written in accordance with the corresponding pg docs. Second I am lacking comments that describe the fields in each struct, where the data is coming from and where it is used.

In general my impression of this (and perhaps some of the related code I have seen coming my way previously) is that you write code that happens to be correct and do the right thing -- which is a feat, I'll grant you the compliment happily --, but without much respect to the future you (or people like the future you) who will not remember how this code was written and why it looks the way it looks. 

tl;dr this area of the product could use some more comments.

And another high-level technical remark: there may be value sometime in caching the content of the virtual table for the duration of a transaction, or until a schema change. I'm not sure what to do with this thought though, feel free to ignore or file for later.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-:-KU3qYD5rTVaGUH_Evtf:brhll64)*

**Comment by nvanbenschoten** [comment]
Your point is fair. Most of this code is declarative and straightforward when looking at the PG docs which is why this isn't heavily documented, but you're right that these aren't actually linked anywhere from here. I've gone back in and added links above each table. I've also added a bit more explanation to the places where the population logic gets more complicated, or where the code deviates from pointing a field in a descriptor at a column in a virtual table. 

I'm not sure which structs you're referring to when you say "Second I am lacking comments that describe the fields in each struct, where the data is coming from and where it is used."

With regard to your second question, I'm not too concerned with the performance of these virtual schemas, especially if improving it means complicating the logic. If this proves to be an issue, we can certainly reconsider.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-:-KU3vvInrbb_TXLLutpQ:b-bofaee)*

**Comment by knz** [comment]
I think this new version is moving in the right direction. At least now there is something the reviewer can double check! Thanks for making this extra step.

---

Reviewed 2 of 4 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/information_schema.go, line 466 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5msz77bDCJnEzSbRE:-KU5msz9QuXI_ZT7tkS_:bd0idho) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/information_schema.go#L466)):*
> ```Go
> // forEachTableDescWithTableLookup acts like forEachTableDesc, except it also provides a
> // tableLookupFn when calling fn to allow callers to lookup fetched table descriptors
> // on demand. This is important for callers dealing with objects like foreign keys.
> ```


Extend this last sentence: ", where the metadata for each object must be augmented by looking at the referenced table."
(Is that correct? I'm not sure by reading the code.)

---

*[pkg/sql/pg_catalog.go, line 702 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5p0g_3ZzcnQ4PJpPO:-KU5p0gaRv6mCcBvNG7o:bj2o8lo) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/pg_catalog.go#L702)):*
> ```Go
> 
> // oidHasher provides a consistent hashing mechanism for object identifiers in
> // pg_catalog tables, allowing for reliable joins across tables.
> ```

Is this documented anywhere in the pg docs? If so please provide a link.

---

*[pkg/sql/pg_catalog.go, line 705 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5ogM0WYuzDaShmrGx:-KU5ogM0WYuzDaShmrGy:bqk28ay) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/pg_catalog.go#L705)):*
> ```Go
> //
> // The type has a few layers of methods:
> // - write<go_type> methods write concreate types to the underlying running hash
> ```

concrete

---

*[pkg/sql/pg_catalog.go, line 713 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5odOZZ3h_c1nwKUat:-KU5odO_PCns9oGJ7k8l:b5mysmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/pg_catalog.go#L713)):*
> ```Go
> //   object identifier for the provided database object. This object identifier will
> //   be returned as a *parser.DInt, and the running hash will be reset. These are the
> //   only methods that are part of the oidHasher's external facing interface
> ```

Missing final period at the end of each bullet point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-:-KU5p9Jgy91XTtj87C12:b-d8ix0m)*

**Comment by nvanbenschoten** [comment]
Review status: 2 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/information_schema.go, line 466 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5msz77bDCJnEzSbRE:-KUJMqpV2wE9wam0dj1h:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/information_schema.go#L466)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> 
> Extend this last sentence: ", where the metadata for each object must be augmented by looking at the referenced table."
> (Is that correct? I'm not sure by reading the code.)
</details>
Done.

---

*[pkg/sql/pg_catalog.go, line 702 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5p0g_3ZzcnQ4PJpPO:-KUJPKkFp2IPcQHsyfwL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/pg_catalog.go#L702)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Is this documented anywhere in the pg docs? If so please provide a link.
</details>
Done.

---

*[pkg/sql/pg_catalog.go, line 705 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5ogM0WYuzDaShmrGx:-KUJPrqUo3KJ9pQ25HbM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/pg_catalog.go#L705)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> concrete
</details>
Done.

---

*[pkg/sql/pg_catalog.go, line 713 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991#-KU5odOZZ3h_c1nwKUat:-KUJPu8YpRYeliinFKeX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c70c37581b1e901ec63bc29efc2d17d1fd9c28d/pkg/sql/pg_catalog.go#L713)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Missing final period at the end of each bullet point.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9991)*

File: pkg/sql/information_schema.go [hunk 1]
```diff
@@ -552,6 +588,25 @@ func forEachColumnInTable(
...
 }
 
+func forEachForeignKeyInTable(
+	table *sqlbase.TableDescriptor,
+	fn func(
+		*sqlbase.IndexDescriptor,
+		*sqlbase.ForeignKeyReference,
+	) error,
+) error {
+	for i := range table.Indexes {
```
Comment (dt) [inline pkg/sql/information_schema.go]: I'm a little sad that we're duplicating this logic again -- the goal of `TableDescriptor.GetConstraintInfo` was to return all the info needed to write something like this and avoid duplication

Comment (nvanbenschoten) [inline pkg/sql/information_schema.go]: The issue is that `GetConstraintInfo` refetches all of the referenced table descriptors for each FK. That's fine if we only want to get the constraint info for a single table, but we already fetched all descriptors here, so we really want to avoid the quadratic blowup. 

If that alone was the issue, it could probably be fixed with some parameterized method to get table descriptors from an ID, but I'm also not sure that `ConstraintDetail` is structured enough for this use case. I would need to add a few new fields to the struct (like `FKColumnIDs` and `IndexColumnIDs`), most of which would be specific to only one constraint type. I would also need to add references to index and table descriptors to the structure. Does this seem like the right approach?

Comment (dt) [inline pkg/sql/information_schema.go]: Hm, it's hard for me to say with confidence without looking at something concrete, but generally, yeah, I think I'd argue in favor of adding more parameterization to a single, common `GetConstraintInfo` and having it return more and more structured data than having a wider proliferation of simpler but overlapping code -- previously enforcement of things like name uniqueness looked up one set of constraints while renaming another, etc which made me really want to centralize the logic that traversed a desc to get *the* set of constraints.

Comment (nvanbenschoten) [inline pkg/sql/information_schema.go]: @dt I just added a new commit that uses `GetConstraintInfo`, which I'll squash if you like it. It was a little intrusive into `GetConstraintInfo` but it actually makes the `pg_constraint` stuff more readable, as it's now easier to see what differs between constraint kinds.

**Review by dt — APPROVED** [review]
